boards/BMS/Src/Io/adc_driver.c

```diff
@@ -0,0 +1,178 @@
+// buffer.
+//
+// The following enum is used to index into raw_adc2_values, which means it must
+// be ordered in ascending ranks. If we were writing an enum for the earlier
+// example, it would look like:
+//
+// enum
+// {
+//     CHANNEL_4, // Rank 1
+//     CHANNEL_7, // Rank 2
+//     CHANNEL_2, // Rank 3
+//     NUM_ADC_CHANNELS,
+// };
+typedef enum
+{
+    RANK_ADC1_CHANNEL_3 = 0U,
+    NUM_ADC1_CHANNELS,
+    RANK_ADC2_CHANNEL_1 = 0U,
+    RANK_ADC2_CHANNEL_3,
+    NUM_ADC2_CHANNELS,
+} AdcRank_E;
+
+typedef struct
+{
+    AdcInstance_E adc_instance; // Specify which ADC the channel corresponds to
+    uint8_t       rank;         // Specifies the rank of the ADC channel
+    float         scale;        // The scale used to convert the raw ADC result to a "real" result
+    float         offset;       // An optional offset field used while converting a raw ADC result to a "real" result
+} AdcConfig_S;
+
+extern ADC_HandleTypeDef hadc1;
+extern ADC_HandleTypeDef hadc2;
+extern TIM_HandleTypeDef htim3;
+
+// Buffers to store raw ADC values
+static uint16_t raw_adc1_buffer[NUM_ADC1_CHANNELS] = { 0U };
+static uint16_t raw_adc2_buffer[NUM_ADC2_CHANNELS] = { 0U };
+
+// Buffers to store converted ADC values
+static float    adc1_voltages[NUM_ADC1_CHANNELS]   = { 0U };
+static float    adc2_voltages[NUM_ADC2_CHANNELS]   = { 0U };
+
+static AdcConfig_S adc_cfg[NUM_ADC_CHANNELS] =
+{
+    // ADC instance 1 configuration
+    [TS_VOLTAGE] =
+    {
+        .adc_instance = ADC_INSTANCE_1,
+        .rank = RANK_ADC1_CHANNEL_3,
+        .scale = TS_VOLTAGE_SCALE,
+        .offset = TS_VOLTAGE_OFFSET_SCALED,
+    },
+
+    // ADC instance 2 configuration
+    [HIGH_RES_TS_CURRENT_SENSE] =
+    {
+        .adc_instance = ADC_INSTANCE_2,
+        .rank = RANK_ADC2_CHANNEL_1,
+        .scale = HIGH_RES_TS_CURRENT_SCALE,
+        .offset = HIGH_RES_TS_CURRENT_OFFSET_SCALED,
+    },
+    [LOW_RES_TS_CURRENT_SENSE] =
+    {
+        .adc_instance = ADC_INSTANCE_2,
+        .rank = RANK_ADC2_CHANNEL_3,
+        .scale = LOW_RES_TS_CURRENT_SCALE,
+        .offset = LOW_RES_TS_CURRENT_OFFSET_SCALED,
+    },
+};
+
+void HAL_ADC_ConvCpltCallback(ADC_HandleTypeDef *hadc)
+{
+    if (hadc->Instance == ADC1)
+    {
+        for (uint8_t channel = TS_VOLTAGE; channel < HIGH_RES_TS_CURRENT_SENSE; channel++)
+        {
+            const uint8_t adc_rank = adc_cfg[channel].rank;
+            adc1_voltages[adc_rank] =
+                GET_CONVERTED_ADC_VALUE(raw_adc1_buffer[channel], adc_cfg[channel].scale, adc_cfg[channel].offset);
+        }
+    }
+    else if (hadc->Instance == ADC2)
+    {
+        for (uint8_t channel = HIGH_RES_TS_CURRENT_SENSE; channel < NUM_ADC_CHANNELS; channel++)
+        {
+            const uint8_t adc_rank = adc_cfg[channel].rank;
+            adc2_voltages[adc_rank] =
+                GET_CONVERTED_ADC_VALUE(raw_adc2_buffer[channel], adc_cfg[channel].scale, adc_cfg[channel].offset);
+        }
+    }
+}
+
+void adc_driver_init(void)
```

Comment:
We should initialize the ADC->DMA->mem pipeline within the scope of this file to avoid exposing pointers unnecessarily

Reply:
I can also pass in the ADC handles into this function. Pretty indifferent. We will likely need to stick with what we're doing right now if we use Modules (functions need to follow template of `void x_init(void)`)